hazelcast/src/main/java/com/hazelcast/transaction/impl/TransactionContextImpl.java

```diff
@@ -79,6 +81,18 @@ public void rollbackTransaction() {
         transaction.rollback();
     }
 
+    @Override
+    public void suspendTransaction() {
+        transaction.set(transaction.cast(TransactionImpl.class).new SuspendedTransactionImpl());
+    }
+
+    @Override
+    public void resumeTransaction() {
+        SuspendedTransactionImpl suspendedTransaction = transaction.cast(SuspendedTransactionImpl.class);
```

Comment:
Is it supported to resume transaction in another thread? If so, what will happen with the `ThreadLocal` in the `TransactionImpl`?

Reply:
> Is it supported to resume transaction in another thread?

yes, this implementation assumes that it's possible to resume transaction from another thread (see [Transaction Suspension](https://docs.oracle.com/cd/E23095_01/Platform.93/ATGProgGuide/html/s1204transactionsuspension01.html)). The thing is that JTA spec have no restrictions on this, also the JBossJTA implementation allow to do so (see [2.3. Suspending and Resuming a Transaction
](https://access.redhat.com/documentation/en-US/JBoss_Enterprise_Web_Platform/5/html/Transactions_JTA_Development_Guide/form-Transactions_JTA_Programmers_Guide-The_API-Suspending_and_resuming_a_transaction.html)).

>  If so, what will happen with the ThreadLocal in the TransactionImpl?

the `ThreadLocal` value becomes `null` on `suspend` and on resume it will be set to caller thread, which can be different from `suspend` caller.